packages/cli/test/integration.js

```diff
@@ -3675,3 +3675,79 @@ test('[vc link] should support the `--project` flag', async t => {
     formatOutput(output)
   );
 });
+
+test('vercel.json projectSettings should be applied to deployment', async t => {
+  const directory = path.join(__dirname, 'fixtures/unit/vercel-json-overrides');
+
+  const deployment = await execa(binaryPath, [
+    directory,
+    ...defaultArgs,
+    '--public',
+    '--confirm',
+  ]);
+
+  t.is(
+    deployment.exitCode,
+    0,
+    formatOutput({
+      stderr: deployment.stderr,
+      stdout: deployment.stdout,
+    })
+  );
+
+  const page = await fetch(deployment.stdout);
+  const text = await page.text();
+  t.is(text, `Hello, World\n`);
+});
+
+test('invalid vercel.json projectSettings should result in a deployment error', async t => {
+  const directory = path.join(
+    __dirname,
+    'fixtures/unit/vercel-json-overrides-error'
+  );
+
+  async function testDeployment() {
+    try {
+      await execa(binaryPath, [
+        directory,
+        ...defaultArgs,
+        '--public',
+        '--confirm',
+      ]);
+      t.fail('Deployment should fail');
+    } catch ({ exitCode, stderr, stdout }) {
+      t.is(
+        exitCode,
+        1,
+        formatOutput({
+          stderr: stderr,
+          stdout: stdout,
+        })
+      );
+    }
+  }
+
+  const vercelJsonPath = path.join(directory, 'vercel.json');
+  const vercelJsonOriginal = await readFile(vercelJsonPath, 'utf8');
+  const vercelConfig = JSON.parse(vercelJsonOriginal);
+
+  // Pass a command that does not exist
+  vercelConfig['projectSettings'] = {
+    command_does_not_exist: 'echo "foo bar"',
+  };
+
+  await writeFile(vercelJsonPath, JSON.stringify(vercelConfig));
+
+  await testDeployment();
+
+  // Pass a disallowed command
+  vercelConfig['projectSettings'] = {
+    autoExposeSystemEnvs: true,
+  };
+
+  await writeFile(vercelJsonPath, JSON.stringify(vercelConfig));
+
+  await testDeployment();
+
+  await writeFile(vercelJsonPath, vercelJsonOriginal);
```

Comment:
We also need a test that creates a new deployment without custom project settings, then modifies the project settings for a second deployment via vercel.json